Improving Mesh in parsers

ndaelman-hu opened the pull request.

ndaelman-hu committed:
Fix conversion from `int64` to `complex128` in the wannier parser

ndaelman-hu committed:
Fix conversion from `int64` to `complex128` in the wannier parser

JosePizarro3 commented:
Could you please add a very simple assertion for Wannier90 regarding the k_mesh.points?

ndaelman-hu committed:
Implement flag for `pytest` interaction with VS code (#81)

for more information, see https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-FAIR/-/issues/1298

Co-authored-by: nathan.daelman@physik.hu-berlin.de <ndaelman@physik.hu-berlin.de>

ndaelman-hu committed:
Connect k mesh extraction for FHIaims + extend testing
Add k offset for FHIaims (no testing data available)

ndaelman-hu committed:
Add FHIaims test

ndaelman-hu committed:
Merge branch 'update_mesh' of github.com:nomad-coe/electronic-parsers into update_mesh

ndaelman-hu committed:
Connect k mesh to `exciting` + add testing

ndaelman-hu committed:
Add check for GW and BSE cases

ndaelman-hu committed:
Connect k mesh + add test

ndaelman-hu committed:
Connect k mesh for QE + add tests

ndaelman-hu committed:
Update k mesh for `VASP` + extend test

ndaelman-hu committed:
Add annotation to FHIaims parser

ndaelman-hu committed:
Apply name change in `FrequencyMesh.value` to `FrequencyMesh.points`

ndaelman-hu committed:
Enforce `MEnum` type of `KMesh.sampling_method`

ndaelman-hu committed:
Correct `KMesh.grid` extraction in Octopus parser

JosePizarro3 commented on electronicparsers/abinit/parser.py:
```diff
@@ -1090,8 +1090,8 @@ def parse_gw(self):
             freqs = frequency_values * ureg.eV
-            values = freqs[:, 0] + freqs[:, 1] * 1j
-            sec_freq_mesh = FrequencyMesh(n_points=len(freqs), values=values)
+            freq_points = freqs[:, 0] + freqs[:, 1] * 1j
```
restore this, dealt with in #80

ndaelman-hu replied:
If it is implemented in #80, then it will get solved when merging.
Are you ready with that PR?

JosePizarro3 replied:
No, you are including further changes here distinct from #80. Please @ndaelman-hu undo these changes. Otherwise it is going to be impossible to track down and we are going to overlook them.....

JosePizarro3 replied:
Please, don't resolve these conversations until they are really resolved by me.

JosePizarro3 commented on electronicparsers/abinit/parser.py:
```diff
@@ -1090,8 +1090,8 @@ def parse_gw(self):
-            values = freqs[:, 0] + freqs[:, 1] * 1j
-            sec_freq_mesh = FrequencyMesh(n_points=len(freqs), values=values)
+            freq_points = freqs[:, 0] + freqs[:, 1] * 1j
+            sec_freq_mesh = FrequencyMesh(n_points=len(freqs), points=freq_points)  # TODO: check
```
restore this, dealt with in #80

JosePizarro3 commented on electronicparsers/exciting/parser.py:
```diff
@@ -735,7 +735,7 @@ def str_to_energy_dict(val_in):
                 'lattice_vectors_reciprocal',
                 r'Reciprocal lattice vectors\s*[\(cartesian\)]*\s*:\s*([\-0-9\.\s]+)\n',
-                str_operation=str_to_array, unit=1 / ureg.bohr, repeats=False, convert=False),
+                str_operation=str_to_array, unit=1 / ureg.bohr, repeats=False, convert=False)
```
delete jump to next line so ] falls in this one.

JosePizarro3 commented on electronicparsers/exciting/parser.py:
```diff
@@ -1502,7 +1502,8 @@ def _parse_input_xs(self, sec_method):
         sec_method.x_exciting_xs_number_of_empty_states = self.input_xml_parser.get(
             'xs/nempty', 5)
         sec_method.x_exciting_xs_ngridq = self.input_xml_parser.get('xs/ngridq', [1, 1, 1])
-        sec_method.x_exciting_xs_ngridk = self.input_xml_parser.get('xs/ngridk', [1, 1, 1])
```
this is creating 2 KMesh sections, once here and once below. I'd say we can keep `x_exciting_xs_ngridk` and delete the next lines after this where you are creating KMesh and asigning the grid.

ndaelman-hu replied:
> this is creating 2 KMesh sections, once here and once below

**Could you pls clarify?** If you mean in the output, it shouldn't, since the section is not repeated. Indeed, I couldn't find any double sections in the tests, `x_exciting_xs_ngridk` included. **Do you have a parsed example output to help see where it is doubled?**
Moreover, I thought that GW or BSE create a new entry?

> I'd say we can keep `x_exciting_xs_ngridk` and delete the next lines after this where you are creating KMesh and asigning the grid.

The point of this PR is to apply the `KMesh` mapping, which is lacking in most parsers (extraction, as here, is mostly present). So I would rather cut out `x_exciting_xs_ngridk`. I didn't touch it in the excited state section, since you preferred to update it yourself. That's fine.

JosePizarro3 replied:
With this I meant that you are creating in this method a KMesh, and then in line 1795-1796 creating another KMesh and populating it using `x_exciting_xs_ngridk`, for which probably you are overwritting KMesh...

There are files in the main installation from BSE for exciting that you can try.

But @ndaelman-hu once more, please read what you are changing: here you are modifying parsing from the xs input file... This `ngridk` only applies in the group named `xs`...

ndaelman-hu replied:
Set back. To avoid future misinterpretations, I suggest that you add a description to `_parse_input_xs()` stating that it covers BSE specifically.

JosePizarro3 replied:
This is implicit on the `xs`, and not only for BSE but also for TDDFT.

ndaelman-hu replied:
Well, that there's this addition just proves my point...
**Implicit is fine, explicit is better.** Documentation is a great tool for long-term maintenance.

JosePizarro3 replied:
yeah, you are right; I will start including docs for each method as far as I modify the parsers

JosePizarro3 commented on electronicparsers/exciting/parser.py:
```diff
@@ -1913,12 +1914,12 @@ def parse_gw(self):
         freqmax = sec_gw.x_exciting_freqgrid.x_exciting_freqmax
         freqmin = sec_gw.x_exciting_freqgrid.x_exciting_freqmin
-        values = [freqmin + i * (freqmax - freqmin) / n_freqs for i in range(n_freqs)] * ureg.hartree
+        freq_points = [freqmin + i * (freqmax - freqmin) / n_freqs for i in range(n_freqs)] * ureg.hartree
```
restore this, dealt with in #80

JosePizarro3 commented on electronicparsers/exciting/parser.py:
```diff
@@ -1913,12 +1914,12 @@ def parse_gw(self):
         smearing = sec_gw.x_exciting_freqgrid.x_exciting_eta if sec_gw.x_exciting_qdepw == 'sum' else None
         sec_freq_mesh = FrequencyMesh(
-            type=sec_gw.x_exciting_freqgrid.x_exciting_fgrid,
+            # type=sec_gw.x_exciting_freqgrid.x_exciting_fgrid,  # TODO: map into enum
```
restore this, dealt with in #80

JosePizarro3 commented on electronicparsers/exciting/parser.py:
```diff
@@ -1913,12 +1914,12 @@ def parse_gw(self):
+            # type=sec_gw.x_exciting_freqgrid.x_exciting_fgrid,  # TODO: map into enum
             n_points=n_freqs,
-            values=values,
+            points=freq_points,  # TODO: check
```
restore this, dealt with in #80

JosePizarro3 commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -977,7 +977,7 @@ def parse_gw(self):
         # KMesh
         sec_k_mesh = sec_method.m_create(KMesh)
-        sec_k_mesh.grid = self.out_parser.get('k_grid', [1, 1, 1])
+        sec_k_mesh.grid = self.out_parser.get('k_grid', [1, 1, 1])  # AIMS does not have a guaranteed PBC, this default should be omitted
```
change to None by default, i.e., delete `[1, 1, 1]` lines

JosePizarro3 commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -986,13 +986,13 @@ def parse_gw(self):
         frequency_data = self.out_parser.get('frequency_data', [])
         if len(frequency_data) > 0:
-            values = np.array(frequency_data)[:, 1] * ureg.hartree
+            freq_points = np.array(frequency_data)[:, 1] * ureg.hartree
```
restore this, dealt with in #80

JosePizarro3 commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -986,13 +986,13 @@ def parse_gw(self):
+            freq_points = np.array(frequency_data)[:, 1] * ureg.hartree
         else:
-            values = None
+            freq_points = None
```
restore this, dealt with in #80

JosePizarro3 commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -986,13 +986,13 @@ def parse_gw(self):
+            freq_points = None
         sec_freq_mesh = FrequencyMesh(
-            type=self.out_parser.get('freq_grid_type'),
+            # type=self.out_parser.get('freq_grid_type'),  # TODO: set default value for enum
```
restore this, dealt with in #80

JosePizarro3 commented on electronicparsers/fhiaims/parser.py:
```diff
@@ -986,13 +986,13 @@ def parse_gw(self):
+            # type=self.out_parser.get('freq_grid_type'),  # TODO: set default value for enum
             n_points=self.out_parser.get('n_freq', 100),
-            values=values)
+            points=freq_points)
```
restore this, dealt with in #80

JosePizarro3 commented on electronicparsers/vasp/parser.py:
```diff
@@ -950,6 +950,14 @@ def get_incar_out(self):
     def n_calculations(self):
         return self.parser.n_calculations
 
+
```
too many blank lines

JosePizarro3 commented on tests/test_abinitparser.py:
```diff
@@ -129,7 +129,7 @@ def test_gw(parser):
     assert sec_gw.screening.n_empty_states == sec_gw.n_empty_states
     assert sec_gw.frequency_mesh.n_points == 2
-    assert sec_gw.frequency_mesh.values[-1].to('eV').magnitude == approx(31.855950000000004j)
+    assert sec_gw.frequency_mesh.points[-1].to('eV').magnitude == approx(31.855950000000004j)  # TODO: check
```
restore this, dealt with in #80

JosePizarro3 commented on electronicparsers/vasp/parser.py:
```diff
@@ -1217,18 +1225,15 @@ def parse_incar(parameters, target):
                 except Exception:
                     self.logger.warning('Error setting metainfo', data=dict(key=key))
+        if sec_k_mesh.points is None:
+            sec_k_mesh.points = [[0.] * 3]
```
not a big fan of this typing; can you please use `[0.0, 0.0, 0.0]`?

ndaelman-hu replied:
Any better reason than aesthetics?

JosePizarro3 replied:
Not really, just to keep some sort of consistency in the parsers, but tbh is the same for me.

JosePizarro3 commented on tests/test_excitingparser.py:
```diff
@@ -51,6 +51,8 @@ def test_gs(parser):
     assert sec_run.program.version == 'CARBON'
 
     sec_method = sec_run.method[0]
+    assert list(sec_method.k_mesh.grid) == [6] * 3
```
`assert (sec_method.k_mesh.grid == np.array([6, 6, 6])).all()`

JosePizarro3 commented on tests/test_excitingparser.py:
```diff
@@ -51,6 +51,8 @@ def test_gs(parser):
 
     sec_method = sec_run.method[0]
+    assert list(sec_method.k_mesh.grid) == [6] * 3
+    assert list(sec_method.k_mesh.offset) == [.0] * 3
```
`assert (sec_method.k_mesh.offset== np.array([0.0, 0.0, 0.0])).all()`

JosePizarro3 commented on tests/test_excitingparser.py:
```diff
@@ -89,7 +91,11 @@ def test_strucopt(parser):
-    sec_sccs = archive.run[0].calculation
+    sec_run = archive.run[0]
+    sec_method = sec_run.method[0]
+    assert list(sec_method.k_mesh.grid) == [6] * 3
```
`assert (sec_method.k_mesh.grid == np.array([6, 6, 6])).all()`

JosePizarro3 commented on tests/test_excitingparser.py:
```diff
@@ -89,7 +91,11 @@ def test_strucopt(parser):
+    sec_run = archive.run[0]
+    sec_method = sec_run.method[0]
+    assert list(sec_method.k_mesh.grid) == [6] * 3
+    assert list(sec_method.k_mesh.offset) == [.0] * 3
```
`assert (sec_method.k_mesh.offset == np.array([0.0, 0.0, 0.0])).all()`

JosePizarro3 commented on tests/test_excitingparser.py:
```diff
@@ -105,9 +111,14 @@ def test_dos_spinpol(parser):
+    sec_scc = sec_run.calculation[0]
     sec_dos = sec_scc.dos_electronic[0]
 
+    assert list(sec_method.k_mesh.grid) == [8] * 3
```
and here

JosePizarro3 commented on tests/test_excitingparser.py:
```diff
@@ -105,9 +111,14 @@ def test_dos_spinpol(parser):
     sec_dos = sec_scc.dos_electronic[0]
 
+    assert list(sec_method.k_mesh.grid) == [8] * 3
+    assert list(sec_method.k_mesh.offset) == [.0] * 3
```
and here?

JosePizarro3 commented on tests/test_excitingparser.py:
```diff
@@ -127,7 +138,12 @@ def test_xs_tddft(parser):
-    sec_sccs = archive.run[0].calculation
+    sec_run = archive.run[0]
+    sec_method = sec_run.method[0]
+    assert list(sec_method.k_mesh.grid) == [12] * 3
```
and here

JosePizarro3 commented on tests/test_excitingparser.py:
```diff
@@ -127,7 +138,12 @@ def test_xs_tddft(parser):
+    sec_run = archive.run[0]
+    sec_method = sec_run.method[0]
+    assert list(sec_method.k_mesh.grid) == [12] * 3
+    assert list(sec_method.k_mesh.offset) == [0.097, 0.273, 0.493]
```
and here

JosePizarro3 commented on tests/test_excitingparser.py:
```diff
@@ -166,9 +182,9 @@ def test_gw(silicon_gw):
     assert sec_gw.screening.n_empty_states == sec_gw.n_empty_states
     assert (sec_gw.screening.k_mesh.grid == np.array([2, 2, 2])).all()
-    assert sec_gw.frequency_mesh.type == 'gauleg2'
+    # assert sec_gw.frequency_mesh.type == 'gauleg2'
```
restore this, dealt with in #80

JosePizarro3 commented on tests/test_excitingparser.py:
```diff
@@ -166,9 +182,9 @@ def test_gw(silicon_gw):
+    # assert sec_gw.frequency_mesh.type == 'gauleg2'
     assert sec_gw.frequency_mesh.n_points == 32
-    assert sec_gw.frequency_mesh.values[4].to('hartree').magnitude == approx(0.125 + 0j)
+    assert sec_gw.frequency_mesh.points[4].to('hartree').magnitude == approx(0.125 + 0j)  # TODO: check
```
restore this, dealt with in https://github.com/nomad-coe/electronic-parsers/pull/80

JosePizarro3 commented on tests/test_excitingparser.py:
```diff
@@ -250,8 +266,10 @@ def test_hybrids(parser):
-    method = archive.run[-1].method[0]
-    assert method.dft.xc_functional.hybrid[0].name == 'HYB_GGA_XC_HSE03'
+    sec_method = archive.run[-1].method[0]
+    assert list(sec_method.k_mesh.grid) == [6, 6, 4]
```
here too, (...).all()

JosePizarro3 commented on tests/test_excitingparser.py:
```diff
@@ -250,8 +266,10 @@ def test_hybrids(parser):
-    assert method.dft.xc_functional.hybrid[0].name == 'HYB_GGA_XC_HSE03'
+    sec_method = archive.run[-1].method[0]
+    assert list(sec_method.k_mesh.grid) == [6, 6, 4]
+    assert list(sec_method.k_mesh.offset) == [0.] * 3
```
here too, (...).all()

JosePizarro3 commented on tests/test_fhiaimsparser.py:
```diff
@@ -64,6 +64,7 @@ def test_scf_spinpol(parser):
 
     assert len(archive.run[0].method) == 1
     sec_method = archive.run[0].method[0]
+    assert list(sec_method.k_mesh.grid) == [16] * 3
```
and here, more verbose and use (...).all()

JosePizarro3 commented on tests/test_fhiaimsparser.py:
```diff
@@ -103,6 +104,7 @@ def test_geomopt(parser):
 
     sec_methods = archive.run[0].method
     assert len(sec_methods) == 1
+    assert list(sec_methods[0].k_mesh.grid) == [8] * 3
```
and here

JosePizarro3 commented on tests/test_fhiaimsparser.py:
```diff
@@ -121,6 +123,9 @@ def test_band_spinpol(parser):
     sec_scc = archive.run[0].calculation[0]
 
+    sec_method = archive.run[0].method[0]
+    assert list(sec_method.k_mesh.grid) == [16] * 3
```
I think you got my point by now 😄

ndaelman-hu replied:
Yeah, no need to repeat it at each statement, but thank you for the effort.

JosePizarro3 commented on tests/test_fhiaimsparser.py:
```diff
@@ -198,6 +203,9 @@ def test_dos(parser):
     parser.parse('tests/data/fhiaims/ClNa_dos/ClNa_dos.out', archive, None)
 
+    sec_method = archive.run[0].method[0]
+    assert list(sec_method.k_mesh.grid) == [10] * 3
```
here

JosePizarro3 commented on tests/test_fhiaimsparser.py:
```diff
@@ -220,6 +228,9 @@ def test_md(parser):
     parser.parse('tests/data/fhiaims/HO_md/H2O_periodic_MD.out', archive, None)
 
+    sec_method = archive.run[0].method[0]
+    assert list(sec_method.k_mesh.grid) == [2] * 3
```
here

JosePizarro3 commented on tests/test_fhiaimsparser.py:
```diff
@@ -253,7 +264,10 @@ def test_hybrid(parser):
 
-    sec_xc_functional = archive.run[0].method[0].dft.xc_functional
+    sec_method = archive.run[0].method[0]
+    assert list(sec_method.k_mesh.grid) == [12] * 3
```
more verbose please

ndaelman-hu replied:
what do you mean with _more verbose_?

JosePizarro3 commented on tests/test_fhiaimsparser.py:
```diff
@@ -264,7 +278,10 @@ def test_dftu(parser):
 
-    sec_hubb = archive.run[-1].method[0].atom_parameters[0].hubbard_kanamori_model
+    sec_method = archive.run[0].method[0]
+    assert list(sec_method.k_mesh.grid) == [1] * 3
```
here

JosePizarro3 commented on tests/test_fhiaimsparser.py:
```diff
@@ -291,8 +308,8 @@ def test_gw(parser):
     assert (sec_method[-1].gw.q_mesh.grid == np.array([1, 1, 1])).all()
-    assert sec_method[-1].gw.frequency_mesh.n_points == len(sec_method[-1].gw.frequency_mesh.values)
-    assert sec_method[-1].gw.frequency_mesh.values[-1].to('hartree').magnitude == approx(3571.4288641158605 + 0j)
+    assert sec_method[-1].gw.frequency_mesh.n_points == len(sec_method[-1].gw.frequency_mesh.points)
```
restore this, dealt with in #80

JosePizarro3 commented on tests/test_fhiaimsparser.py:
```diff
@@ -338,7 +355,9 @@ def test_gw_bands(parser):
     assert sec_run.system[-1].atoms.labels == ['Si', 'Si']
-    assert sec_run.method[-1].gw.type == 'G0W0'
+    sec_method = sec_run.method[-1]
+    assert list(sec_method.k_mesh.grid) == [8] * 3
```
and here

JosePizarro3 commented on tests/test_quantumespressoparser.py:
```diff
@@ -124,7 +127,10 @@ def test_md(parser):
-    sec_sccs = archive.run[0].calculation
+    sec_run = archive.run[0]
+    sec_method = sec_run.method[0]
+    assert len(sec_method.k_mesh.points) == 1
```
Does k_mesh make any sense for MD?

ndaelman-hu replied:
It's run in an ab initio code with pbc and it has a k mesh. You're right, in that in somewhat detracts from the main point.
Tbh, I'm fine either way, but I'm not going to push just for ambivalent stuff.

What I mostly want from a review is feedback on design, logical, implementation or performance errors, including for use cases that I didn't predict. Style matters in as much as it violates PEP8 or the native pycodestyle.

JosePizarro3 reviewed (commented):
Done with the review. Mainly being consistent when coding and restoring the changes dealt with in #80

ndaelman-hu committed:
Applied reviewer's comments

JosePizarro3 reviewed (changes requested):
Ok, I have several comments here that I want you to deal with, because this is being too time consuming.

This is a TODO:
- [x] Please, revert the changes you did for FrequencyMesh. There is more than what you did, and it is going to be impossible to track the differences.
- [x] Also change back the exciting `xs/ngridk` KMesh parsing, as right now I am guessing it is incorrect and producing errors that you are not able to debug. I can deal with the repeated quantities in a separate issue, as it is a good practice.

And please:
- Can you let the comments unresolved? I am working on checking the github directly. Everytime you resolve, I have to open and close each one to see what is going on... This is very cumbersome.
- Style also goes beyond PEP8 and pycodestyle. I understand some students and developers see this as a secondary thing, but keeping a structure in the code makes it more readable for later people. Now someone will arrive after you, will look into VASP or QE and will think: wtf, how are they so different? Not to speak of the other 30 ab initio parsers...
- In any case, I admit these things are kind of subjective. My comments are mostly line by line, similar to reviewing a thesis or a paper (that's why might look kind of repetitive), and are mostly suggestions, so no need to accept them 🙂

JosePizarro3 committed:
Frequency and time meshes (#80)

Update FrequencyMesh and TimeMesh
_____________________________________________

* Fixed ABINIT parser and test for new Mesh base class

* Fixed exciting parser and test for new Mesh base class

Fix units on frequency_mesh.points

* Fixed FHIaims parser and test for new Mesh base class

* Fixed VASP parser and test for new Mesh base class

* Moved dmft.n_tau and dmft.n_matsubaras to FrequencyMesh and TimeMesh in method for w2dynamics

Fixed w2dynamics test

* Moved dmft.n_tau and n_matsubaras to FrequencyMesh and TimeMesh in method for solid_dmft

Fixed solid_dmft test

* Adding frequency grid type MEnum to FHIaims and exciting

* Improved regex for freq_grid_type

* Fixed Wannier90 parser for complex kpoints

Fixed FHIaims test

* Added condition for regex on k_grid to not create empty KMesh and QMesh

* Solved issues when rebasing into update_mesh

ndaelman-hu committed:
Update k-mesh parsing in exciting

ndaelman-hu committed:
Add `k_mesh.points` test to Wannier90

JosePizarro3 reviewed (approved):
Everything looks nice 👍 Thanks! 

Merge it whenever you want.
